lib/ts/filterChoicesUtils.ts

```diff
@@ -0,0 +1,111 @@
+import { Answers, IPromptFilterStrategy, PromptListChoice, UIBuildType, UserFlagsRaw } from "./types.js";
+
+const getIsValueSupported = (value: string, supportedValues: string[]) => {
+    return supportedValues.includes(value);
+};
+
+const filterFrontendByUiType: IPromptFilterStrategy = {
+    validValues: ["react"],
+    filterChoices(choices, answers) {
+        if (answers.ui === UIBuildType.PRE_BUILT) {
+            return choices;
+        }
+        return choices.filter((choice) => getIsValueSupported(choice.value, this.validValues));
+    },
+    validateUserArguments(userArguments) {
+        if (!userArguments.frontend) return true;
+        return (
+            userArguments.ui === UIBuildType.PRE_BUILT || getIsValueSupported(userArguments.frontend, this.validValues)
+        );
+    },
+};
+
+const filterRecipeByUiType: IPromptFilterStrategy = {
+    validValues: ["emailpassword", "thirdpartyemailpassword"],
+    filterChoices(choices, answers) {
+        if (answers.ui === UIBuildType.PRE_BUILT) {
+            return choices;
+        }
+        return choices.filter((choice) => getIsValueSupported(choice.value, this.validValues));
+    },
+    validateUserArguments(userArguments) {
+        if (!userArguments.recipe) return true;
+        return (
+            userArguments.ui === UIBuildType.PRE_BUILT || getIsValueSupported(userArguments.recipe, this.validValues)
+        );
+    },
+};
+
+export const FILTER_CHOICES_STRATEGY = {
+    filterFrontendByUiType,
+    filterRecipeByUiType,
+};
+
+/**
+ * Filters a list of choices based on the provided strategy or strategies.
+ *
+ * @param choices - An array of `PromptListChoice` objects or a promise that resolves to such an array.
+ * @param answers - An object containing the answers to previous prompts.
+ * @param strategy - An optional filtering strategy or an array of strategies. This can be one of the predefined
+ *                   `FILTER_CHOICES_STRATEGY` values, an array of such values, or a custom function that takes an
+ *                   array of choices and returns a promise that resolves to a filtered array of choices.
+ * @returns A promise that resolves to the filtered array of `PromptListChoice` objects.
+ */
+export const filterChoices = async (
+    choices: PromptListChoice[] | Promise<PromptListChoice[]>,
+    answers: Answers,
+    strategy?:
+        | IPromptFilterStrategy
+        | IPromptFilterStrategy[]
```

Comment:
No need for array type

Reply:
The code is updated with the requested changes, hence resolving this comment.